File: product-brain/src/product_brain/bot/worker.py
from __future__ import annotations

import time
import traceback
import uuid
from datetime import datetime, timezone
from typing import Callable, Optional

from ..adapters import get as get_adapter
from ..config import Config
from ..models import TicketDraft
from ..planner import run_command
from .audit import AuditEntry, AuditLog
from .comment import build_comment, content_hash, locate_existing
from .cooldown import in_quiet_hours, within_cooldown
from .queue import Job, Queue


def _process_job(job: Job, config: Config, queue: Queue, audit: AuditLog) -> None:
    adapter = get_adapter(config.pm_adapter, config)
    now = datetime.now(timezone.utc)

    is_auto = job.trigger.startswith("status:") or job.trigger.startswith("auto:")
    is_manual = job.trigger.startswith("comment:") or job.requester != "system"

    last = audit.last_for_ticket(job.ticket_id, job.command)
    if is_auto:
        if last and within_cooldown(last.timestamp, now, config.bot.cooldown_hours):
            queue.complete(job.id)
            return
        if in_quiet_hours(now, config.bot.quiet_hours_utc):
            queue.complete(job.id)
            return

    body, summary, model, cost = run_command(config, job.command, job.ticket_id, args=job.payload.get("args", ""))

    h = content_hash(body)
    if last and last.input_hash == h:
        queue.complete(job.id)
        audit.append(AuditEntry(
            id=str(uuid.uuid4()), timestamp=time.time(), trigger=job.trigger,
            ticket_id=job.ticket_id, command=job.command, requester=job.requester,
            input_hash=h, output_summary="(skipped: unchanged)", model=model, cost_usd=0.0,
        ))
        return

    change_note: Optional[str] = None
    if last and last.input_hash and last.input_hash != h:
        change_note = "Inputs changed since last run."

    run_id = str(uuid.uuid4())
    full_body = build_comment(
        body=body,
        command=job.command,
        trigger=job.trigger,
        requester=job.requester,
        run_id=run_id,
        change_note=change_note,
        run_at=now,
    )

    if job.command == "draft-tickets":
        adapter.post_comment(job.ticket_id, full_body)
        for draft in _extract_drafts(body):
            adapter.create_ticket(TicketDraft(
                title=draft["title"],
                description=draft["description"],
                parent_id=job.ticket_id,
                status=config.bot.draft_status,
            ))
    else:
        comments = adapter.list_comments(job.ticket_id)
        existing = locate_existing(comments, job.command)
        if existing:
            adapter.edit_comment(job.ticket_id, existing.id, full_body)
        else:
            adapter.post_comment(job.ticket_id, full_body)

    audit.append(AuditEntry(
        id=run_id, timestamp=time.time(), trigger=job.trigger,
        ticket_id=job.ticket_id, command=job.command, requester=job.requester,
        input_hash=h, output_summary=summary, model=model, cost_usd=cost,
    ))
    queue.complete(job.id)


def _extract_drafts(body: str) -> list[dict]:
    out: list[dict] = []
    in_section = False
    current: dict = {}
    lines = body.splitlines()
    for line in lines:
        if line.strip().startswith("## Draft sub-tickets"):
            in_section = True
            continue
        if not in_section:
            continue
        if line.startswith("### "):
            if current:
                out.append(current)
            current = {"title": line[4:].strip(), "description": ""}
        elif current:
            current["description"] += line + "\n"
    if current:
        out.append(current)
    return out


def run_worker(config: Config, poll_seconds: float = 2.0):
    queue = Queue(config.queue.path)
    audit = AuditLog(config.audit.path)
    while True:
        job = queue.claim_next()
        if not job:
            time.sleep(poll_seconds)
            continue
        try:
            _process_job(job, config, queue, audit)
        except Exception as e:
            queue.fail(job.id, f"{e}\n{traceback.format_exc()}")
            audit.append(AuditEntry(
                id=str(uuid.uuid4()), timestamp=time.time(), trigger=job.trigger,
                ticket_id=job.ticket_id, command=job.command, requester=job.requester,
                error=str(e),
            ))

File: product-brain/src/bot/worker.ts
import { execFileSync } from "node:child_process";
import { randomUUID } from "node:crypto";
import { setTimeout as sleep } from "node:timers/promises";
import { getPmAdapter } from "../adapters/index.js";
import { backfillRepo, type BackfillResult } from "../backfill/run.js";
import type { Config } from "../config.js";
import { runCommand } from "../planner.js";
import { AuditLog, type AuditEntry } from "./audit.js";
import { buildComment, contentHash, locateExisting } from "./comment.js";
import { inQuietHours, withinCooldown } from "./cooldown.js";
import { Queue, type Job } from "./queue.js";

function commitBrainRepo(config: Config, repoName: string, summary: BackfillResult): void {
  const brain = config.brainRoot;
  try {
    execFileSync("git", ["-C", brain, "add", `repos/${repoName}`], { stdio: "ignore" });
  } catch {
    // nothing to add yet
  }
  const msg = `sync(${repoName}): ${summary.written} updated, ${summary.created} new`;
  try {
    execFileSync("git", ["-C", brain, "commit", "-m", msg], { stdio: "ignore" });
  } catch {
    return;
  }
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      execFileSync("git", ["-C", brain, "pull", "--rebase"], { stdio: "ignore" });
      execFileSync("git", ["-C", brain, "push"], { stdio: "ignore" });
      return;
    } catch {
      // retry with backoff
      const delay = (2 ** attempt) * 1000;
      const start = Date.now();
      while (Date.now() - start < delay) {
        // sync sleep is fine inside worker; queue serializes jobs
      }
    }
  }
}

async function processSourceMerge(job: Job, config: Config, queue: Queue, audit: AuditLog): Promise<void> {
  const repoName = String(job.payload["repo"] ?? "");
  const sinceSha = job.payload["since_sha"] as string | undefined;
  const summary = await backfillRepo(config, repoName, { since: sinceSha });
  commitBrainRepo(config, repoName, summary);
  audit.append({
    id: randomUUID(),
    timestamp: Date.now() / 1000,
    trigger: job.trigger,
    ticketId: job.ticketId,
    command: job.command,
    requester: job.requester,
    outputSummary: JSON.stringify(summary),
    model: "",
    costUsd: 0,
  });
  queue.complete(job.id);
}

function extractDrafts(body: string): Array<{ title: string; description: string }> {
  const out: Array<{ title: string; description: string }> = [];
  let inSection = false;
  let current: { title: string; description: string } | null = null;
  for (const line of body.split("\n")) {
    if (line.trim().startsWith("## Draft sub-tickets")) {
      inSection = true;
      continue;
    }
    if (!inSection) continue;
    if (line.startsWith("### ")) {
      if (current) out.push(current);
      current = { title: line.slice(4).trim(), description: "" };
    } else if (current) {
      current.description += line + "\n";
    }
  }
  if (current) out.push(current);
  return out;
}

async function processJob(job: Job, config: Config, queue: Queue, audit: AuditLog): Promise<void> {
  if (job.command === "source-merge") {
    return processSourceMerge(job, config, queue, audit);
  }

  const adapter = getPmAdapter(config.pmAdapter, config);
  const now = new Date();

  const isAuto = job.trigger.startsWith("status:") || job.trigger.startsWith("auto:");
  const last = audit.lastForTicket(job.ticketId, job.command);

  if (isAuto) {
    if (last && withinCooldown(last.timestamp, now, config.bot.cooldown_hours)) {
      queue.complete(job.id);
      return;
    }
    if (inQuietHours(now, config.bot.quiet_hours_utc)) {
      queue.complete(job.id);
      return;
    }
  }

  const { body, summary, model, cost } = await runCommand(
    config,
    job.command,
    job.ticketId,
    String(job.payload["args"] ?? ""),
  );

  const h = contentHash(body);
  if (last && last.inputHash === h) {
    queue.complete(job.id);
    audit.append({
      id: randomUUID(),
      timestamp: Date.now() / 1000,
      trigger: job.trigger,
      ticketId: job.ticketId,
      command: job.command,
      requester: job.requester,
      inputHash: h,
      outputSummary: "(skipped: unchanged)",
      model,
      costUsd: 0,
    });
    return;
  }

  const changeNote = last && last.inputHash && last.inputHash !== h
    ? "Inputs changed since last run."
    : undefined;

  const runId = randomUUID();
  const fullBody = buildComment({
    body,
    command: job.command,
    trigger: job.trigger,
    requester: job.requester,
    runId,
    changeNote,
    runAt: now,
  });

  if (job.command === "draft-tickets") {
    await adapter.postComment(job.ticketId, fullBody);
    for (const draft of extractDrafts(body)) {
      await adapter.createTicket({
        title: draft.title,
        description: draft.description,
        parentId: job.ticketId,
        status: config.bot.draft_status,
      });
    }
  } else {
    const comments = await adapter.listComments(job.ticketId);
    const existing = locateExisting(comments, job.command);
    if (existing) await adapter.editComment(job.ticketId, existing.id, fullBody);
    else await adapter.postComment(job.ticketId, fullBody);
  }

  audit.append({
    id: runId,
    timestamp: Date.now() / 1000,
    trigger: job.trigger,
    ticketId: job.ticketId,
    command: job.command,
    requester: job.requester,
    inputHash: h,
    outputSummary: summary,
    model,
    costUsd: cost,
  });
  queue.complete(job.id);
}

export async function runWorker(config: Config, pollSeconds = 2): Promise<never> {
  const queue = new Queue(config.queue.path);
  const audit = new AuditLog(config.audit.path);
  while (true) {
    const job = queue.claimNext();
    if (!job) {
      await sleep(pollSeconds * 1000);
      continue;
    }
    try {
      await processJob(job, config, queue, audit);
    } catch (e) {
      const err = e as Error;
      queue.fail(job.id, `${err.message}\n${err.stack ?? ""}`);
      audit.append({
        id: randomUUID(),
        timestamp: Date.now() / 1000,
        trigger: job.trigger,
        ticketId: job.ticketId,
        command: job.command,
        requester: job.requester,
        error: err.message,
      });
    }
  }
}
